Mean Amplitude Deviation metric

Asanto32 opened the pull request:
Resolves #132 

Adding the mean amplitude deviation metric, alternative to enmo.

Asanto32 commented:
Might wait to merge the bugfix PR first

frey-perez reviewed (commented):
Don't have much to add here except adding a little more testing for non-zero expected values. Don't know how necessary that is, given that the function is relatively simple.

ReinderVosDeWael reviewed (commented):
How do I use this?

Asanto32 commented on src/wristpy/processing/metrics.py:
```diff
@@ -58,6 +58,44 @@ def angle_relative_to_horizontal(
+
+    An alternative to ENMO to quantify the intensity of physical activity.
+    It is calculated as the mean of the absolute difference between acceleration values
+    and the mean of those acceleration values in a 5s window.
```
The paper explicitly said 5s but it can be a variable that defaults to 5s?

Asanto32 commented:
Ready for re-review @ReinderVosDeWael  @frey-perez 

Changed how the thresholds argument was being handled and moved some of the error handling around. Let me know what you think. I was having trouble with the None type or List of floats using nargs = 3 (I tried `nargs= '?' ` but no luck), hence the new `_none_or_float_list` function.

ReinderVosDeWael commented on src/wristpy/core/cli.py:
```diff
@@ -53,18 +53,24 @@ def parse_arguments(args: Optional[List[str]] = None) -> argparse.Namespace:
+    parser.add_argument(
+        "-a",
+        "--activity_metric",
+        type=str,
```
IIRC you can set this to `lambda s: s.lower()` to accept any case but force it to always be lower case.

Asanto32 replied:
Should I make this change for the calibrator as well? Note I have switched the activity metric choices to lower case,  "enmo" and "mad"

Asanto32 replied:
Made the change in both cases to handle any mixed case entries/typos in user input

ReinderVosDeWael commented on src/wristpy/core/orchestrator.py:
```diff
@@ -98,12 +98,13 @@ def format_nonwear_data(
         Literal["ggir", "gradient"],
     ] = "gradient",
     epoch_length: Union[int, None] = 5,
+    activity_metric: Literal["ENMO", "MAD"] = "ENMO",
```
`ggir` in the other argument is all lower case. Pick either all upper or all lower for acronyms, don't mix the two.

Asanto32 replied:
Switched to all lower case, this goes for the argparse options in the CLI as well

ReinderVosDeWael commented on src/wristpy/processing/metrics.py:
```diff
@@ -58,6 +58,53 @@ def angle_relative_to_horizontal(
+    ).set_sorted("time")
+
+    mad_df = (
+        mad_lf.group_by_dynamic(index_column="time", every=f"{epoch_length}s")
```
Could convert to nanoseconds and set epoch_length to a float? That'd allow for epoch lengths of e.g. 3.5seconds

ReinderVosDeWael commented on src/wristpy/core/cli.py:
```diff
@@ -122,21 +131,52 @@ def main(
     )
+
+
+def _none_or_float_list(value: str) -> Optional[List[float]]:
```
Yeah.. I'm starting to like the idea of checking out tyro. There's nothing wrong with this function, but this would be so much easier if we can just define a Pydantic model.

Asanto32 replied:
Potentially a new issue to be added to our task list
